File: repeat_py/279.py
class Solution:
    def numSquares(self, n: int) -> int:
        mem = [0] * (n + 1)
        sqRoot = 1
        for i in range(1, n + 1):
            if i == sqRoot * sqRoot:
                mem[i] = 1
                sqRoot += 1
            else:
                minVal = i
                for j in range(1, sqRoot):
                    minVal = min(minVal, mem[j * j] + mem[i - j * j])
                mem[i] = minVal
        return mem[n]

class SolutionConcise:
    def numSquares(self, n: int) -> int:
        mem = [0] * (n + 1)
        for i in range(1, n + 1):
            minVal = n
            j = 1
            while j * j <= i:
                # min = min(minVal, mem[j * j] + mem[i - j * j]); *** every mem[j*j] is actually '1'
                minVal = min(minVal, 1 + mem[i - j * j])
                j += 1
            mem[i] = minVal
        return mem[n]


Solution().numSquares(2)

File: problems/310/medium/dp/279.java
class Solution {
    public int numSquares(int n) {
        int[] mem = new int[n + 1];
        int sq = 1;
        for (int i = 1; i < n + 1; i++) {
            if (i == sq * sq) {
                mem[i] = 1;
                sq++;
            } else {
                int min = n;
                for (int j = 1; j < sq; j++) {
                    min = Math.min(min, mem[j * j] + mem[i - j * j]);
                }
                mem[i] = min;
            }
        }
        return mem[n];
    }
}

class SolutionConcise { // can use squares as a boundaries and start from 0
    public int numSquares(int n) {
        int[] mem = new int[n + 1];
        mem[1] = 1;
        for (int i = 2; i < n + 1; i++) {
            int min = n;
            for (int j = 1; j*j <= i; j++) {
                // min = Math.min(min, mem[j * j] + mem[i - j * j]); *** every mem[j*j] is actually '1'
                min = Math.min(min, 1 + mem[i - j * j]);
            }
            mem[i] = min;
        }
        return mem[n];
    }
}
